### PR TITLE
Fix js compile 2

### PR DESCRIPTION
**Pull request by johnspackman**
Replaces https://github.com/qooxdoo/qooxdoo-compiler/pull/402

### CODE REVIEW

**Comment by cboulanger** [comment]
@johnspackman Is there a reason why you don't `require()` compile.js like a node module, which exports a single function? (This is not a change request, just to understand)

File: docs/configuration/compile.md [hunk 1]
```diff
@@ -379,23 +379,88 @@ If you choose to use the optional web server by running `qx serve`, you can chan
...
+        
+    } else {
+        return { /* ... snip ... */ };
+    }
+)();
+```
+
+Or an example using callbacks:
+```
+function(inputData, cb) {
```
Comment (hkollmann) [inline docs/configuration/compile.md]: inputData should be removed here too,.

**Comment by johnspackman** [comment]
@cboulanger originally my goal was to produce usable error messages (eval is useless in that regard), but using the `vm.Script` also allows the `compile.js` to be run in an isolated context, where only those variables you want to make available to the script are available - this includes global variables.

In [qxl.apiviewer/compile.js](https://github.com/qooxdoo/qxl.apiviewer/blob/master/compile.js#L3) this meant that even the global `qx` was not available and `require("@qooxdoo/framework");` had to be added.

I'll take responsibility for fixing demobrowser etc if this PR is accepted - but also I'm happy to switch to `require("./compile.js")` if that's what everybody else would prefer.  Mostly I wanted compile.js to have real error messages and line numbers!

**Comment by hkollmann** [comment]
Personally i prefer using build in Methods like require instead of external libraries. Much less stuff to maintain …

**Comment by cboulanger** [comment]
I am happy with leaving things as they are - but as the solution is a bit unusual (at least for me), I would explain them a bit more detailed as I have tried above - does that make sense to you, @johnspackman ?

**Comment by johnspackman** [comment]
@hkollmann Personally i prefer using build in Methods like require instead of external libraries. Much less stuff to maintain …

It is a built in library, it's the official node way of running external scripts... https://nodejs.org/dist/latest-v10.x/docs/api/vm.html

**Comment by cboulanger** [comment]
@johnspackman I see! Now I understand the mechanism. I thought the code would be eval'd. Maybe you can add to the documentation that this is how `compile.js` is run?

**Comment by cboulanger** [comment]
Updated my suggestion above to reflect what I learned 😄

**Comment by cboulanger** [comment]
2 approvals, merged PR. Thank you @johnspackman !

**Pull request merged**